File: ios-client/OutfitSuggestor/Utils/AiProgressSteps.swift
//
//  AiProgressSteps.swift
//  OutfitSuggestor
//
//  Staged progress definitions for long-running AI operations.
//

import Foundation
import Combine

enum AiOperationType: Equatable {
    case outfitSuggestion
    case outfitWithPreview
    case wardrobeOutfit
    case wardrobeAnalysis
    case randomHistory
    case pastSuggestions
}

struct AiProgressStep: Identifiable, Equatable {
    let id: String
    let label: String
    let durationMs: Int
}

enum AiProgressSteps {
    static func steps(for operation: AiOperationType) -> [AiProgressStep] {
        switch operation {
        case .outfitSuggestion:
            return [
                AiProgressStep(id: "prepare", label: "Preparing your image", durationMs: 2500),
                AiProgressStep(id: "analyze", label: "Analyzing your item", durationMs: 8000),
                AiProgressStep(id: "match", label: "Matching colors and style", durationMs: 10000),
                AiProgressStep(id: "build", label: "Building outfit recommendation", durationMs: 15000),
            ]
        case .outfitWithPreview:
            return [
                AiProgressStep(id: "prepare", label: "Preparing your image", durationMs: 2500),
                AiProgressStep(id: "analyze", label: "Analyzing your item", durationMs: 8000),
                AiProgressStep(id: "match", label: "Matching colors and style", durationMs: 10000),
                AiProgressStep(id: "build", label: "Building outfit recommendation", durationMs: 12000),
                AiProgressStep(id: "preview", label: "Generating preview", durationMs: 20000),
            ]
        case .wardrobeOutfit:
            return [
                AiProgressStep(id: "scan", label: "Scanning your wardrobe", durationMs: 4000),
                AiProgressStep(id: "match", label: "Matching colors and style", durationMs: 8000),
                AiProgressStep(id: "build", label: "Building outfit recommendation", durationMs: 12000),
            ]
        case .wardrobeAnalysis:
            return [
                AiProgressStep(id: "review", label: "Reviewing your wardrobe", durationMs: 5000),
                AiProgressStep(id: "gaps", label: "Identifying gaps", durationMs: 10000),
                AiProgressStep(id: "recommend", label: "Preparing recommendations", durationMs: 15000),
            ]
        case .randomHistory:
            return [
                AiProgressStep(id: "fetch", label: "Loading your saved looks", durationMs: 2500),
                AiProgressStep(id: "pick", label: "Finding a varied outfit", durationMs: 2000),
                AiProgressStep(id: "prepare", label: "Preparing your look", durationMs: 2000),
            ]
        case .pastSuggestions:
            return [
                AiProgressStep(id: "fetch", label: "Loading your saved looks", durationMs: 2500),
                AiProgressStep(id: "filter", label: "Finding outfits for this item", durationMs: 2000),
                AiProgressStep(id: "prepare", label: "Preparing suggestions", durationMs: 2000),
            ]
        }
    }

    static func title(for operation: AiOperationType) -> String {
        switch operation {
        case .outfitSuggestion, .outfitWithPreview:
            return "Creating your outfit"
        case .wardrobeOutfit:
            return "Building from your wardrobe"
        case .wardrobeAnalysis:
            return "Analyzing your wardrobe"
        case .randomHistory:
            return "Picking from your history"
        case .pastSuggestions:
            return "Loading past suggestions"
        }
    }

    static func estimatedTotalSeconds(for operation: AiOperationType) -> Int {
        let totalMs = steps(for: operation).reduce(0) { $0 + $1.durationMs }
        return max(1, totalMs / 1000)
    }

    static func formatDuration(seconds: Int) -> String {
        if seconds < 60 { return "\(seconds)s" }
        let mins = seconds / 60
        let secs = seconds % 60
        return secs > 0 ? "\(mins)m \(secs)s" : "\(mins)m"
    }

    static func stepIndex(for message: String?, steps: [AiProgressStep]) -> Int {
        guard let message else { return 0 }
        let lower = message.lowercased()
        if steps.first?.id == "fetch" {
            if lower.contains("saved look") || lower.contains("loading your saved") || lower.contains("history") {
                return 0
            }
            if steps.count > 1, steps[1].id == "filter" {
                if lower.contains("prepar") {
                    return steps.count - 1
                }
                if lower.contains("filter")
                    || lower.contains("finding outfit")
                    || (lower.contains("for this item") && lower.contains("finding")) {
                    return min(1, steps.count - 1)
                }
                if lower.contains("past") || lower.contains("suggestion") {
                    return 0
                }
            } else if lower.contains("varied") || (lower.contains("pick") && !lower.contains("history")) {
                return min(1, steps.count - 1)
            }
            if lower.contains("preparing your look") || lower.contains("prepar") || lower.contains("scan") {
                return steps.count - 1
            }
        }
        if lower.contains("history") || lower.contains("saved look") { return 0 }
        if lower.contains("varied") || (lower.contains("pick") && !lower.contains("history")) {
            return min(1, steps.count - 1)
        }
        if lower.contains("preparing your look") { return min(2, steps.count - 1) }
        if lower.contains("compress") || lower.contains("prepar") { return 0 }
        if lower.contains("analyz") || lower.contains("review") { return min(1, steps.count - 1) }
        if lower.contains("match") || lower.contains("color") || lower.contains("scan") { return min(2, steps.count - 1) }
        if lower.contains("different") || lower.contains("generating ai") || lower.contains("creating your outfit") {
            return min(3, steps.count - 1)
        }
        if lower.contains("preview") || lower.contains("model") { return steps.count - 1 }
        if lower.contains("premium") || lower.contains("chatgpt") || lower.contains("free rules") {
            return min(1, steps.count - 1)
        }
        if lower.contains("gap") || lower.contains("identif") { return min(1, steps.count - 1) }
        return 0
    }
}

@MainActor
final class StagedAiProgressTracker: ObservableObject {
    @Published private(set) var activeStepIndex = 0
    @Published private(set) var elapsedSeconds = 0

    private var timer: Timer?
    private var sessionStartedAt = Date()
    private var stepStartedAt = Date()
    private var operation: AiOperationType = .outfitSuggestion
    private var steps: [AiProgressStep] = []

    var progressPercent: Int {
        min(95, Int(((Double(activeStepIndex) + 0.35) / Double(max(steps.count, 1))) * 100))
    }

    var estimatedTotalSeconds: Int {
        AiProgressSteps.estimatedTotalSeconds(for: operation)
    }

    var estimatedRemainingSeconds: Int {
        max(0, estimatedTotalSeconds - elapsedSeconds)
    }

    var showSlowHint: Bool {
        elapsedSeconds >= 45
    }

    func start(operation: AiOperationType, message: String?) {
        stop()
        self.operation = operation
        self.steps = AiProgressSteps.steps(for: operation)
        sessionStartedAt = Date()
        stepStartedAt = Date()
        activeStepIndex = AiProgressSteps.stepIndex(for: message, steps: steps)
        elapsedSeconds = 0

        timer = Timer.scheduledTimer(withTimeInterval: 0.5, repeats: true) { [weak self] _ in
            Task { @MainActor in
                self?.tick()
            }
        }
    }

    func updateMessage(_ message: String?) {
        let messageIndex = AiProgressSteps.stepIndex(for: message, steps: steps)
        activeStepIndex = max(activeStepIndex, messageIndex)
    }

    func stop() {
        timer?.invalidate()
        timer = nil
        activeStepIndex = 0
        elapsedSeconds = 0
    }

    private func tick() {
        elapsedSeconds = Int(Date().timeIntervalSince(sessionStartedAt))
        guard activeStepIndex < steps.count else { return }
        let currentStep = steps[activeStepIndex]
        let stepElapsedMs = Int(Date().timeIntervalSince(stepStartedAt) * 1000)
        if stepElapsedMs >= currentStep.durationMs, activeStepIndex < steps.count - 1 {
            stepStartedAt = Date()
            activeStepIndex += 1
        }
    }
}

File: frontend/src/utils/aiProgressSteps.ts
import { useEffect, useMemo, useRef, useState } from 'react';

export type AiOperationType =
  | 'outfit-suggestion'
  | 'outfit-with-preview'
  | 'wardrobe-outfit'
  | 'wardrobe-analysis'
  | 'random-history'
  | 'past-suggestions';

export interface AiProgressStep {
  id: string;
  label: string;
  /** Typical duration before advancing to the next simulated step. */
  durationMs: number;
}

export const AI_PROGRESS_STEPS: Record<AiOperationType, AiProgressStep[]> = {
  'outfit-suggestion': [
    { id: 'prepare', label: 'Preparing your image', durationMs: 2500 },
    { id: 'analyze', label: 'Analyzing your item', durationMs: 8000 },
    { id: 'match', label: 'Matching colors and style', durationMs: 10000 },
    { id: 'build', label: 'Building outfit recommendation', durationMs: 15000 },
  ],
  'outfit-with-preview': [
    { id: 'prepare', label: 'Preparing your image', durationMs: 2500 },
    { id: 'analyze', label: 'Analyzing your item', durationMs: 8000 },
    { id: 'match', label: 'Matching colors and style', durationMs: 10000 },
    { id: 'build', label: 'Building outfit recommendation', durationMs: 12000 },
    { id: 'preview', label: 'Generating preview', durationMs: 20000 },
  ],
  'wardrobe-outfit': [
    { id: 'scan', label: 'Scanning your wardrobe', durationMs: 4000 },
    { id: 'match', label: 'Matching colors and style', durationMs: 8000 },
    { id: 'build', label: 'Building outfit recommendation', durationMs: 12000 },
  ],
  'wardrobe-analysis': [
    { id: 'review', label: 'Reviewing your wardrobe', durationMs: 5000 },
    { id: 'gaps', label: 'Identifying gaps', durationMs: 10000 },
    { id: 'recommend', label: 'Preparing recommendations', durationMs: 15000 },
  ],
  'random-history': [
    { id: 'fetch', label: 'Loading your saved looks', durationMs: 2500 },
    { id: 'pick', label: 'Finding a varied outfit', durationMs: 2000 },
    { id: 'prepare', label: 'Preparing your look', durationMs: 2000 },
  ],
  'past-suggestions': [
    { id: 'fetch', label: 'Loading your saved looks', durationMs: 2500 },
    { id: 'filter', label: 'Finding outfits for this item', durationMs: 2000 },
    { id: 'prepare', label: 'Preparing suggestions', durationMs: 2000 },
  ],
};

export function getEstimatedDurationMs(operationType: AiOperationType): number {
  return AI_PROGRESS_STEPS[operationType].reduce((sum, step) => sum + step.durationMs, 0);
}

export function formatDuration(seconds: number): string {
  if (seconds < 60) return `${seconds}s`;
  const mins = Math.floor(seconds / 60);
  const secs = seconds % 60;
  return secs > 0 ? `${mins}m ${secs}s` : `${mins}m`;
}

export function resolveStepFromMessage(message: string | null | undefined, steps: AiProgressStep[]): number {
  if (!message) return 0;
  const lower = message.toLowerCase();
  if (steps[0]?.id === 'fetch') {
    if (lower.includes('saved look') || lower.includes('loading your saved') || lower.includes('history')) {
      return 0;
    }
    if (steps[1]?.id === 'filter') {
      if (lower.includes('prepar')) {
        return steps.length - 1;
      }
      if (
        lower.includes('filter') ||
        lower.includes('finding outfit') ||
        (lower.includes('for this item') && lower.includes('finding'))
      ) {
        return Math.min(1, steps.length - 1);
      }
      if (lower.includes('past') || lower.includes('suggestion')) {
        return 0;
      }
    } else if (lower.includes('pick') || lower.includes('varied')) {
      return Math.min(1, steps.length - 1);
    }
    if (lower.includes('prepar') || lower.includes('scan')) {
      return steps.length - 1;
    }
  }
  if (lower.includes('compress')) return 0;
  if (lower.includes('analyz')) return 1;
  if (lower.includes('match') || lower.includes('color')) return 2;
  if (lower.includes('different outfit') || lower.includes('generating ai')) return 3;
  if (lower.includes('preview') || lower.includes('model')) return steps.length - 1;
  if (lower.includes('premium') || lower.includes('chatgpt')) return 1;
  if (lower.includes('free rules')) return 1;
  return 0;
}

export interface StagedAiProgress {
  steps: AiProgressStep[];
  activeStepIndex: number;
  elapsedSeconds: number;
  estimatedTotalSeconds: number;
  estimatedRemainingSeconds: number;
  progressPercent: number;
  showSlowHint: boolean;
}

export function useStagedAiProgress(
  isActive: boolean,
  operationType: AiOperationType,
  loadingMessage?: string | null
): StagedAiProgress {
  const steps = AI_PROGRESS_STEPS[operationType];
  const [activeStepIndex, setActiveStepIndex] = useState(0);
  const [elapsedSeconds, setElapsedSeconds] = useState(0);
  const stepStartedAtRef = useRef<number>(Date.now());
  const sessionStartedAtRef = useRef<number>(Date.now());

  useEffect(() => {
    if (!isActive) {
      setActiveStepIndex(0);
      setElapsedSeconds(0);
      return;
    }

    const now = Date.now();
    sessionStartedAtRef.current = now;
    stepStartedAtRef.current = now;
    setActiveStepIndex(resolveStepFromMessage(loadingMessage, steps));
    setElapsedSeconds(0);
  }, [isActive, operationType]); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    if (!isActive) return;

    const messageIndex = resolveStepFromMessage(loadingMessage, steps);
    setActiveStepIndex((current) => Math.max(current, messageIndex));
  }, [isActive, loadingMessage, steps]);

  useEffect(() => {
    if (!isActive) return;

    const tick = window.setInterval(() => {
      const elapsed = Math.floor((Date.now() - sessionStartedAtRef.current) / 1000);
      setElapsedSeconds(elapsed);

      const stepElapsed = Date.now() - stepStartedAtRef.current;
      const currentStep = steps[activeStepIndex];
      if (currentStep && stepElapsed >= currentStep.durationMs && activeStepIndex < steps.length - 1) {
        stepStartedAtRef.current = Date.now();
        setActiveStepIndex((index) => Math.min(index + 1, steps.length - 1));
      }
    }, 500);

    return () => window.clearInterval(tick);
  }, [isActive, activeStepIndex, steps]);

  return useMemo(() => {
    const estimatedTotalSeconds = Math.round(getEstimatedDurationMs(operationType) / 1000);
    const estimatedRemainingSeconds = Math.max(0, estimatedTotalSeconds - elapsedSeconds);
    const completedSteps = activeStepIndex;
    const progressPercent = Math.min(
      95,
      Math.round(((completedSteps + 0.35) / steps.length) * 100)
    );

    return {
      steps,
      activeStepIndex,
      elapsedSeconds,
      estimatedTotalSeconds,
      estimatedRemainingSeconds,
      progressPercent,
      showSlowHint: elapsedSeconds >= 45,
    };
  }, [operationType, steps, activeStepIndex, elapsedSeconds]);
}
